### PR TITLE
Update ByteBufferDataChunk.isReleased and ByteBufDataChunk.isReleased to use AtomicBoolean

### LINKED ISSUE #6845: MultiPart IllegalReferenceCountException: refCnt: 0, decrement: 1
## Environment Details
* Helidon Version: 2.6.0
* Helidon SE

----------

Consuming a body part content as an `InputStream` in a different threads triggers the following exception:

```java
io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
	at io.netty.common@4.1.86.Final/io.netty.util.internal.ReferenceCountUpdater.toLiveRealRefCnt(ReferenceCountUpdater.java:83)
	at io.netty.common@4.1.86.Final/io.netty.util.internal.ReferenceCountUpdater.release(ReferenceCountUpdater.java:147)
	at io.netty.buffer@4.1.86.Final/io.netty.buffer.AbstractReferenceCountedByteBuf.release(AbstractReferenceCountedByteBuf.java:101)
	at io.helidon.common.http@2.6.0/io.helidon.common.http.ByteBufferDataChunk.release(ByteBufferDataChunk.java:86)
	at io.helidon.media.multipart@2.6.0/io.helidon.media.multipart.BodyPartChunk.release(BodyPartChunk.java:49)
	at io.helidon.media.common@2.6.0/io.helidon.media.common.DataChunkInputStream.releaseChunk(DataChunkInputStream.java:86)
	at io.helidon.media.common@2.6.0/io.helidon.media.common.DataChunkInputStream.read(DataChunkInputStream.java:164)
	at java.base/java.nio.channels.Channels$ReadableByteChannelImpl.read(Channels.java:387)
	at com.oracle.cx.industries.edk.integrations.base.http@23.10.0.0-SNAPSHOT/com.oracle.cx.industries.edk.integrations.base.http.internal.MultipartSpokeReplySubscriber.lambda$onNext$3(MultipartSpokeReplySubscriber.java:107)
	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:718)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
	at io.helidon.common.reactive@2.6.0/io.helidon.common.reactive.SingleToFuture.onNext(SingleToFuture.java:67)
	at io.helidon.common.reactive@2.6.0/io.helidon.common.reactive.SingleObserveOn$ObserveOnSubscriber.run(SingleObserveOn.java:127)
	at io.helidon.common.context@2.6.0/io.helidon.common.context.Contexts.runInContext(Contexts.java:117)
	at io.helidon.common.context@2.6.0/io.helidon.common.context.ContextAwareExecutorImpl.lambda$wrap$7(ContextAwareExecutorImpl.java:154)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

Reproducer:

```java
ExecutorService executor = ThreadPoolSupplier.builder()
                                                   .corePoolSize(5)
                                                   .build()
                                                   .get();

MultiPartSupport multiPartSupport = MultiPartSupport.create();
WebClient webClient = WebClient.builder()
                               .baseUri("http://localhost:8080")
                               .addMediaSupport(multiPartSupport)
                               .build();

WebServer.builder()
         .port(8080)
         .addMediaSupport(multiPartSupport)
         .routing(Routing.builder()
                         .get("/test", (req, res) -> {
                             final var capacity = 21 << 20;
                             final var buffer = ByteBuffer.allocate(capacity);
                             final byte[] contents = new byte[capacity - 1];
                             Arrays.fill(contents, (byte) 0x20);
                             buffer.put(contents);
                             buffer.put((byte) 0x00);
                             buffer.flip();
                                 /*
                                    now we have a buffer with 21MB capacity, all filled with spaces except the last
                                    byte that is the NUL or \0 byte to terminate the string
                                  */
                             res.send(WriteableMultiPart
                                     .builder()
                                     .bodyPart(WriteableBodyPart
                                             .builder()
                                             .publisher(Single.just(DataChunk.create(buffer)))
                                             .headers(WriteableBodyPartHeaders
                                                     .builder()
                                                     .contentType(TEXT_PLAIN)
                                                     .contentDisposition(ContentDisposition
                                                             .builder()
                                                             .name("too-big")
                                                             .build())
                                                     .build())
                                             .build())
                                     .build()); // we actually don't need payload here
                         })
                         .get("/", (req, res) ->
                                 webClient.get()
                                          .path("test")
                                          .accept(MediaType.MULTIPART_FORM_DATA)
                                          .submit()
                                          .flatMap(response -> response.content().asStream(ReadableBodyPart.class))
                                          .subscribe(new Flow.Subscriber<>() {

                                              Flow.Subscription subscription;

                                              @Override
                                              public void onSubscribe(Flow.Subscription subscription) {
                                                  this.subscription = subscription;
                                                  subscription.request(1L);
                                              }

                                              @Override
                                              public void onNext(ReadableBodyPart part) {
                                                  part.content()
                                                      .as(InputStream.class)
                                                      .observeOn(executor)
                                                      .onError(req::next)
                                                      .forSingle(is -> {
                                                          try {
                                                              byte[] bytes = is.readAllBytes();
                                                              subscription.request(1L);
                                                          } catch (Throwable e) {
                                                              req.next(e);
                                                          }
                                                      })
                                                      .exceptionally(res::send);
                                                  subscription.request(1L);
                                              }

                                              @Override
                                              public void onError(Throwable throwable) {
                                                  throwable.printStackTrace(System.out);
                                                  res.send("OK");
                                              }

                                              @Override
                                              public void onComplete() {
                                                  res.send("OK");
                                              }
                                          })))
         .build()
         .start();
```

### PR DESCRIPTION
**Pull request by romain-grecourt**
Update `ByteBufferDataChunk.isReleased` and `ByteBufDataChunk.isReleased` to use an AtomicBoolean.
This prevents race conditions that may call the release callback more than once.

Fixes #6845

### CODE REVIEW

File: tests/integration/se-gh-6845/src/main/java/io/helidon/tests/integration/gh6845/Main.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    /**
+     * Creates new {@link io.helidon.webserver.Routing}.
+     *
+     * @return routing configured
+     */
+    private static Routing createRouting() {
+        return Routing.builder()
+                      .get("/", (req, res) -> {
+                          byte[] contents = new byte[21 << 20 - 1];
```
Comment (danielkec) [inline tests/integration/se-gh-6845/src/main/java/io/helidon/tests/integration/gh6845/Main.java:95]: Comment explaining the magic number for the future could be appreciated some day :smile:

Comment (romain-grecourt) [reply tests/integration/se-gh-6845/src/main/java/io/helidon/tests/integration/gh6845/Main.java:95]: `21 << 20` is 21 MiB in bytes.

```java
1 << 10 // 2 ^ 10 = 1024
1 << 20 // 2 ^ 20 = 1024 * 1024
21 << 20 // 21 * 2 ^ 20 = 21 * 1024 * 1024
```

TBH I took it as-is from the person that reported the initial issue. I'm not sure what the `-1` is for, maybe this was meant for less than `21 MiB` but in that case it should have been `(21 << 20) - 1`

File: common/http/src/main/java/io/helidon/common/http/ByteBufferDataChunk.java [hunk 2]
```diff
@@ -29,7 +30,7 @@ final class ByteBufferDataChunk implements DataChunk {
     private final boolean flush;
     private final boolean readOnly;
     private final Runnable releaseCallback;
-    private boolean isReleased = false;
+    private final AtomicBoolean isReleased = new AtomicBoolean();
```
Comment (ljnelson) [inline common/http/src/main/java/io/helidon/common/http/ByteBufferDataChunk.java:33]: You might consider using `VarHandle` instead, so you aren't lugging around a full `AtomicBoolean` inside every single `DataChunk`, just one static field and a `volatile boolean` instead.

File: common/http/src/main/java/io/helidon/common/http/ByteBufferDataChunk.java [hunk 3]
```diff
@@ -25,11 +27,21 @@
  */
 final class ByteBufferDataChunk implements DataChunk {
 
+    private static final VarHandle IS_RELEASED;
+
+    static {
+        try {
+            IS_RELEASED = MethodHandles.lookup().findVarHandle(ByteBufferDataChunk.class, "isReleased", boolean.class);
+        } catch (NoSuchFieldException | IllegalAccessException e) {
+            throw new RuntimeException(e);
```
Comment (ljnelson) [inline common/http/src/main/java/io/helidon/common/http/ByteBufferDataChunk.java:36]: (Or `throw (Error)new ExceptionInInitializerError(e.getMessage()).initCause(e);`. Not a big deal either way.)

File: webserver/webserver/src/main/java/io/helidon/webserver/ByteBufDataChunk.java [hunk 4]
```diff
@@ -32,11 +34,20 @@
  */
 public class ByteBufDataChunk implements DataChunk {
 
+    private static final VarHandle IS_RELEASED;
+
+    static {
+        try {
+            IS_RELEASED = MethodHandles.lookup().findVarHandle(ByteBufDataChunk.class, "isReleased", boolean.class);
+        } catch (NoSuchFieldException | IllegalAccessException e) {
+            throw new RuntimeException(e);
```
Comment (ljnelson) [inline webserver/webserver/src/main/java/io/helidon/webserver/ByteBufDataChunk.java:43]: (Same tiny comment as earlier.)

**Review by ljnelson — APPROVED** [review]
LGTM; restricted my review to the `VarHandle` stuff

**Review by ljnelson — APPROVED** [review]
LGTM for `VarHandle` stuff

File: common/http/src/main/java/io/helidon/common/http/ByteBufferDataChunk.java [hunk 5]
```diff
@@ -25,11 +27,21 @@
...
+            throw (Error) new ExceptionInInitializerError(e.getMessage()).initCause(e);
+        }
+    }
+
     private final ByteBuffer[] byteBuffers;
     private final boolean flush;
     private final boolean readOnly;
     private final Runnable releaseCallback;
-    private boolean isReleased = false;
+    private volatile boolean isReleased;
```
Comment (danielkec) [inline common/http/src/main/java/io/helidon/common/http/ByteBufferDataChunk.java:44]: Ever wondered why is AtomicBoolean using volatile int internally? :smiling_imp:

**Review by danielkec — APPROVED** [review]
To be honest I liked AtomicBoolean variant more, it had same performance with better readability.

**Review by danielkec — CHANGES_REQUESTED** [review]
Please use var handle over int field rather than boolean

**Review by ljnelson — APPROVED** [review]

**Pull request merged**